### PR TITLE
Update tprof to use trace sessions

### PR DESCRIPTION
**Pull request by garazdawi**
This PR is based on #8330 and adds trace session support to tprof.

tprof now has one global server that uses a single trace session and the temporary servers used by `tprof:profile` create their own sessions, thus allowing the global server and ad-hoc profiling to be used together.

I have not fixed the tprof API to allow running multiple named servers as I don't think it is very likely that it will be used like that.

### CODE REVIEW

File: lib/tools/src/tprof.erl [hunk 1]
```diff
@@ -868,36 +856,42 @@ profile(Module, Function, Args, Options) when is_atom(Module), is_atom(Function)
 init(Config) ->
     Type = maps:get(type, Config, call_count),
     false = erlang:process_flag(trap_exit, true), %% need this for reliable terminate/2 call
-    {ok, #tprof_state{type=Type}}.
+    Session = trace:session_create(maps:get(session_name, Config, ?MODULE), self(), []),
```
Comment (max-au) [inline lib/tools/src/tprof.erl:859]: Dialyzer probably does not see that, but `session_name` appears to be what `registered_name` used to stand for before this change. I think this behaviour should be documented.

Now that we have multiple trace sessions support, there is not much sense in the trick original `hprof` did. We can have many `tprof` sessions running, identified by PID.

**Review by max-au — COMMENTED** [review]
I think that `tprof` API should be extended to support multiple trace sessions.

I can imagine it this way:

```
{ok, Tracer1} = tprof:start_link( .... ),
{ok, Tracer2} = tprof:start_link( .... ),
tprof:enable_trace(Tracer1, ...),
...
tprof:collect(Tracer2),
...
```

Default (singleton) tracer may be a registered process, and calls could be just like:
```
collect() ->
    collect(?MODULE).
```
I used that approach in `pg`, and it felt rather natural & convenient.

File: lib/tools/src/tprof.erl (hunk 1 above)
Comment (garazdawi) [reply lib/tools/src/tprof.erl:859]: > Dialyzer probably does not see that, but session_name appears to be what registered_name used to stand for before this change. I think this behaviour should be documented.

The session name does not have to be unique per node, so it is somewhat different from `registered_name`. The tprof server is registered as `tprof` when not using the oneshot API, but if we do use the oneshot we can have as many sessions as we want in parallel.

**Comment by garazdawi** [comment]
> I think that `tprof` API should be extended to support multiple trace sessions.
> 
> I can imagine it this way:
> 
> ```
> {ok, Tracer1} = tprof:start_link( .... ),
> {ok, Tracer2} = tprof:start_link( .... ),
> tprof:enable_trace(Tracer1, ...),
> ...
> tprof:collect(Tracer2),
> ...
> ```
> 
> Default (singleton) tracer may be a registered process, and calls could be just like:
> 
> ```
> collect() ->
>     collect(?MODULE).
> ```
> 
> I used that approach in `pg`, and it felt rather natural & convenient.

The problem is the APIs like `tprof:enable_trace/1,2` when we want to have both a default and optional tracer. The spec becomes something like this:

```erlang
-spec enable_trace(spec()) -> non_neg_integer().
-spec enable_trace(tracer(), spec()) -> non_neg_integer();
                  (spec(), trace_options()) -> non_neg_integer().
-spec enable_trace(tracer(), spec(), trace_options()) -> non_neg_integer().
```

and we try to avoid that type of API overloading if we can. Maybe we can skip `enable_trace(tracer(), spec())` clause and force users to provide the options if they want to use the non-default session. Just like `io:format/1,2,3` do with `device()`.

File: lib/tools/src/tprof.erl (hunk 1 above)
Comment (max-au) [reply lib/tools/src/tprof.erl:859]: I mean, even if it's different from the `registered_name`, it seems worth documenting.

**Comment by max-au** [comment]
> Maybe we can skip enable_trace(tracer(), spec()) clause and force users to provide the options if they want to use the non-default session

Either that, or split `enable_trace` in two different functions:
```
-spec enable_trace(spec()) -> non_neg_integer().
-spec enable_trace(tracer(), spec()) -> non_neg_integer();
                  (spec(), trace_options()) -> non_neg_integer().

-spec enable_trace_options(spec(), trace_options()) -> non_neg_integer().
-spec enable_trace_options(tracer(), spec(), trace_options()) -> non_neg_integer().
```

**Comment by garazdawi** [comment]
Pushed a commit where I added `enable_trace/3` and `disable_trace/3` and session variants to all other things. Also added tests that run (in parallel!) to test all the things.

**Review by max-au — APPROVED** [review]
That's even fancier than I imagined, thanks, I love it!

File: lib/tools/test/tprof_SUITE.erl [hunk 2]
```diff
@@ -396,23 +435,26 @@ hierarchy() ->
...
     ?assert(Traced > 5),
-    ?assert(tprof:set_pattern(code_server, '_', '_') > 5),
+    ?assert(set_pattern(Config, Srv, code_server, '_', '_') > 5),
     _ = code:get_path(), %% makes a call to code_server
     %% disabling all processes tracing should return more than "children of"
-    ?assert(tprof:disable_trace(processes) > Traced),
-    {call_memory, Profile} = tprof:collect(),
-    tprof:stop(),
+    Disabled = disable_trace(Config, Srv, all),
+    ct:log("~p",[{Disabled, Traced}]),
```
Comment (max-au) [inline lib/tools/test/tprof_SUITE.erl:445]: Minor nit, this just looks like a debug printout leftover.

File: lib/tools/test/tprof_SUITE.erl [hunk 3]
```diff
@@ -219,23 +238,35 @@ set_on_spawn(Config) when is_list(Config) ->
             {Pid, MRef} = spawn_monitor(fun () -> lists:seq(1, 32) end),
             receive {'DOWN', MRef, process, Pid, normal} -> done end
         end, #{report => return, set_on_spawn => false, type => call_memory}),
-    {done, ProfileMFA} = tprof:profile(?MODULE, seq, [32], #{report => return, set_on_spawn => false, type => call_memory}),
+    ct:log("~p",[Profile]),
```
Comment (max-au) [inline lib/tools/test/tprof_SUITE.erl:241]: ...and this one as well.

File: lib/tools/test/tprof_SUITE.erl (hunk 2 above)
Comment (garazdawi) [reply lib/tools/test/tprof_SUITE.erl:448]: Yes, I'll remove them once all tests pass.

Comment (max-au) [reply lib/tools/test/tprof_SUITE.erl:469]: I still see some, unless you wanted to keep it for future debugging.

Comment (garazdawi) [reply lib/tools/test/tprof_SUITE.erl:469]: I've left it for now to debug tests... I'm having issues with timeouts on our 200+ core machines.

**Pull request merged**